addons/base_automation/models/base_automation.py

```diff
@@ -280,106 +281,102 @@ def _compute_trg_date_calendar_id(self):
+        self.on_change_field_ids = False
 
-    @api.depends('model_id', 'trigger')
+    @api.depends('trigger')
     def _compute_trigger_field_ids(self):
         for automation in self:
-            domain = [('model_id', '=', automation.model_id.id)]
-            if automation.trigger == 'on_stage_set':
-                domain += [('ttype', '=', 'many2one'), ('name', 'in', ['stage_id', 'x_studio_stage_id'])]
-            elif automation.trigger == 'on_tag_set':
-                domain += [('ttype', '=', 'many2many'), ('name', 'in', ['tag_ids', 'x_studio_tag_ids'])]
-            elif automation.trigger == 'on_priority_set':
-                domain += [('ttype', '=', 'selection'), ('name', 'in', ['priority', 'x_studio_priority'])]
-            elif automation.trigger == 'on_state_set':
-                domain += [('ttype', '=', 'selection'), ('name', 'in', ['state', 'x_studio_state'])]
-            elif automation.trigger == 'on_user_set':
-                domain += [
-                    ('relation', '=', 'res.users'),
-                    ('ttype', 'in', ['many2one', 'many2many']),
-                    ('name', 'in', ['user_id', 'user_ids', 'x_studio_user_id', 'x_studio_user_ids']),
-                ]
-            elif automation.trigger in ['on_archive', 'on_unarchive']:
-                domain += [('ttype', '=', 'boolean'), ('name', 'in', ['active', 'x_active'])]
-            elif automation.trigger == 'on_time_created':
-                domain += [('ttype', '=', 'datetime'), ('name', '=', 'create_date')]
-            elif automation.trigger == 'on_time_updated':
-                domain += [('ttype', '=', 'datetime'), ('name', '=', 'write_date')]
-            else:
-                automation.trigger_field_ids = False
-                continue
-            if automation.model_id.is_mail_thread and automation.trigger in MAIL_TRIGGERS:
-                continue
-
-            automation.trigger_field_ids = self.env['ir.model.fields'].search(domain, limit=1)
+            field = (
+                automation._get_trigger_related_field()
+                if automation.trigger not in TIME_TRIGGERS
+                else False
+            )
+            automation.trigger_field_ids = field
 
-    @api.depends('trigger_field_ids')
+    @api.depends('model_id')
     def _compute_trigger(self):
-        for automation in self:
-            automation.trigger = False if not automation.trigger_field_ids else automation.trigger
+        self.trigger = False
+
+    @api.onchange('filter_domain')
+    def _onchange_filter_domain(self):
+        if self.trigger not in ['on_change', 'on_create_or_write']:
+            return
+        domain_fields = [
+            part[0].split(".")[0]
+            for part in safe_eval.safe_eval(self.filter_domain, self._get_eval_context())
+            if is_leaf(part)
+        ] if self.filter_domain else []
+        fields_to_append = self.env['ir.model.fields'].search([
```

Comment:
we usually use `self.env['ir.model.fields']._get` to retrieve fields (or you can use the `_get_ids` for one lookup but it’s cached anyway)